index.html

```diff
@@ -1823,29 +1835,338 @@ <h5><code>invokeaction</code></h5>
+          Accept: application/json
+          </pre>
+          <p class="rfc2119-assertion"
+          id="profile-5.2.2.2-thing-protocol-binding-actions-queryaction-4">
             If a Web Thing receives an HTTP request following the format
-            above and the Consumer has permission to invoke the corresponding
-            action, then upon successfully invoking the action it MUST send an
-            HTTP response with:
-            </span>
+            above and the Consumer has permission to query the corresponding
+            <code>ActionStatus</code> resource, then upon successfully reading 
+            the status of the action request it MUST send an HTTP response 
+            with:
           </p>
-          <p class="ednote">
-            The response to invoking an action needs to be defined.
-            <span class="rfc2119-assertion" id="profile-5.2.2.1-thing-protocol-binding-actions-invokeaction-4">
-            Given not all actions can be completed within the
-            timeout period of an HTTP response, this may need to include a
-            reference to an action request resource in an action queue (see
-            <a href="https://github.com/w3c/wot-thing-description/issues/302">
-            #302</a>).
-            </span>
+          <ul>
+            <li>Status code set to <code>200</code></li>
+            <li><code>Content-Type</code> header set to <code>application/json
+              </code></li>
+            <li>A body containing an <code>ActionStatus</code> object
+              representing the current status of the action request, serialized 
+            in JSON</li>
+          </ul>
+          <pre class="example">
+            HTTP/1.1 200 OK
+            Content-Type: application/json
+            {
+              "status": "running"
+            }
+          </pre>
+          <p class="rfc2119-assertion" 
+          id="profile-5.2.2.2-thing-protocol-binding-actions-queryaction-5">
+            If the queried action failed to execute, then the 
+            <code>status</code> member of the <code>ActionStatus</code> object 
+            MUST be set to <code>"failed"</code> and the <code>error</code>
+            member may optionally provide additional error information 
+            conforming to the Problem Details format [[RFC7807]].
+          </p>
+          <pre class="example">
+            HTTP/1.1 200 OK
+            Content-Type: application/json
+            {
+              "status": "failed",
+              "error": {
+                "type": "https://mythingserver.com/docs/errors/invalid-level",
+                "title": "Invalid value for level provided",
+                "invalid-params": [
+                  {
+                    "name": "level",
+                    "reason": "Must be a valid number between 0 and 100",
+                  }
+                ]
+              }
+            }
+          </pre>
+        </section>
+
+        <section id="cancelaction">
```

Comment:
I believe the cancelaction operation is underspecified in this proposal:

The current text does not describe the error behavior and status response mechanism, if cancellation is not successful.
It does not consider that cancelling an action could take longer than the request timeout, i.e. the cancel operation would be de facto another asynchronous action with a status URL.

It would be straightforward to impement cancel operations by just using the available mechanism and  perform a regular invocation operation (note that it could be sync or async) depending on the use case. 

Therefore I suggest to eliminate it completely and recommending a naming convention such as ending the action name with "Cancel" for a corresponding cancel operation, e.g.
on a class Teapot:
- brewTea(duration)
- brewTeaCancel()

Reply:
Hi @mlagally, thank you for the feedback.

> The current text does not describe the error behavior and status response mechanism, if cancellation is not successful.

Error behaviour is common to all operations and is defined in section 5.2.4 as I was asked to move it out into a separate section. This does mean that there are no recommendations of specific error codes for specific scenarios, but I'd be happy to add those in a follow-up if you have suggestions.

> It does not consider that cancelling an action could take longer than the request timeout

Do you have a particular use case in mind where cancelling an action might take more than 30 seconds? If this is a problem then one possibility would be to define a special case for this scenario by having the Web Thing respond to a `cancelaction` operation with `202 Accepted` instead of `204 No Content` to indicate that the request has been accepted by not yet acted upon.

> It would be straightforward to impement cancel operations by just using the available mechanism and perform a regular invocation operation (note that it could be sync or async) depending on the use case.
> 
> Therefore I suggest to eliminate it completely and recommending a naming convention such as ending the action name with "Cancel" for a corresponding cancel operation, e.g.
> on a class Teapot:
> 
> * brewTea(duration)
> * brewTeaCancel()

I remember that you suggested this approach before. There are a couple of reasons I don't think we should take that approach in the general case. The first is that it means that every cancellable Action affordance becomes two Action affordances, unnecessarily cluttering the interaction model. But the biggest problem in my view is that it doesn't deal well with multiple ongoing instances of the same action as it's very hard to keep track of which cancel operations correspond with which invoke operations. The point of the action queue pattern is to spawn a dynamic resource for each ongoing action such that individual instances can be identified, queried and cancelled. The approach you have proposed has the effect of creating two separate queues (an invoke queue and a cancel queue) which must then somehow be kept in sync with each other.

I would instead suggest that for the vast majority of cases where cancelling an action doesn't take longer than 30 seconds, the current approach provides a simpler and more effective model. For edge cases where a cancel operation might take longer, developers always have the freedom to make the initial action non-cancellable and define a second action affordance which can reverse the other. For example, a 3D printer could have a "print" operation and a separate "reset" operation which resets its state back to the starting position and has the effect of cancelling any ongoing "print" operations. There's no need to enforce a naming convention for this.

Reply:
We discussed this in the arch call on Sept. 9th:
On a timeout of a cancel operation it should be checked whether the status object is still available, in this case the cancel operation was not successful, since otherwise the action status object would no longer exist.